### PR TITLE
Fix detected memory leaks in deterministic-unit_test_einspline

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This frees owning raw pointers that are currently leaking memory.

This is part of #3083

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- n/a. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- n/a. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
Let us use unique_ptr Deleter to make the code RAII.

File: src/einspline/tests/test_3d.cpp [hunk 1]
```diff
@@ -86,44 +86,48 @@ TEST_CASE("double_3d_periodic","[einspline]")
   bc.lCode = PERIODIC;
   bc.rCode = PERIODIC;
 
-  UBspline_3d_d* s = create_UBspline_3d_d(grid, grid, grid, bc, bc, bc, data);
+  auto s = std::unique_ptr<UBspline_3d_d, void (*)(void*)>{create_UBspline_3d_d(grid, grid, grid, bc, bc, bc, data),
```
Comment (ye-luo) [inline src/einspline/tests/test_3d.cpp:89]: It seems working very well. Could you update all the pointers in test_3d.cpp and test_one.cpp?

**Review by ye-luo — DISMISSED** [review]

**Comment by ye-luo** [comment]
Test this please

File: src/einspline/tests/test_one.cpp [hunk 2]
```diff
@@ -182,35 +178,36 @@ TEST_CASE("multi_cuda_wrapper", "[einspline]")
   BCtype_s xBC;
   xBC.lCode = NATURAL;
   xBC.rCode = NATURAL;
-  multi_UBspline_1d_s* s = create_multi_UBspline_1d_s(x_grid, xBC, 1);
+  auto s    = std::unique_ptr<multi_USspline_1d_s, void (*)(void*)>{create_multi_UBspline_1d_s(x_grid, xBC, 1),
```
Comment (ye-luo) [inline src/einspline/tests/test_one.cpp:181]: UBspline caught by CI.

Comment (quantumsteve) [reply src/einspline/tests/test_one.cpp]: Fixed. Surprised clang 11 built with the wrong pointer type.

File: src/einspline/tests/test_3d.cpp (hunk 1 above)
Comment (quantumsteve) [reply src/einspline/tests/test_3d.cpp:89]: done

File: src/einspline/tests/test_one.cpp (hunk 2 above)
Comment (ye-luo) [reply src/einspline/tests/test_one.cpp]: It is under CUDA build.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Comment by prckent** [comment]
Thanks Steve.

**Pull request merged**